### PR TITLE
Sys shd with posix and node

### PR DESCRIPTION
**Pull request by RobertoAldea**

### CODE REVIEW

File: code/sys_shd/src/system_shared_tool/sys_shd_channels.py [hunk 1]
```diff
@@ -209,3 +209,19 @@ def is_empty(self) -> bool:
...
+    def terminate(self) -> None:
+        """Terminate the queue .
+        """
+        closed = False
+        try:
+            self.close()
+            closed = True
+        except ipc.ExistentialError as err:
+            log.error(f"Trying to close queue {self.name} with error {err}")
+        if closed:
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_channels.py]: Move unlink to first try

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py [hunk 2]
```diff
@@ -19,6 +25,69 @@
 
 #######################             CLASSES              #######################
 
+class SysShdParamsC:
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_common.py]: Rename to SysShdNodeParamsC

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py [hunk 3]
```diff
@@ -19,6 +25,69 @@
...
+    def process_iteration(self) -> None:
+        """Function to be implemented in the inherited class
+        """
+
+    def stop(self) -> None:
+        """
+        Stop the node thread .
+        """
+        log.critical("Stopping node thread.")
+        self.join()
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_common.py]: Add try, except if node is down

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py [hunk 4]
```diff
@@ -35,3 +104,12 @@ def __init__(self, message) -> None:
...
 
 #######################            FUNCTIONS             #######################
+def wait(tic: float, sleep_time: float) -> None:
+    """Wait until the sleep time is reached.
+
+    Args:
+        tic ([float]): [description]
+        sleep_time ([float]): [Period in seconds]
+    """
+    while time()-tic <= sleep_time:
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_common.py]: Remove active wait

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py [hunk 5]
```diff
@@ -19,6 +25,69 @@
...
+        '''
+        Main method executed by the CAN thread. It receive data from EPCs and PLAKs
+        and store it on the corresponding chan.
+        '''
+        log.info("Start running process")
+        while self.working_flag.isSet():
+            try:
+                tic = time()
+                self.process_iteration()
+                wait(tic, self.cycle_period)
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_common.py]: Change to blocking wait

**Review by mariuscrsn — CHANGES_REQUESTED** [review]

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py (hunk 3 above)
Comment (RobertoAldea) [reply code/sys_shd/src/system_shared_tool/sys_shd_common.py]: self.join() doesn´t make sense here as the action can not be done in the current node

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py (hunk 4 above)
Comment (RobertoAldea) [reply code/sys_shd/src/system_shared_tool/sys_shd_common.py]: Done

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py (hunk 5 above)
Comment (RobertoAldea) [reply code/sys_shd/src/system_shared_tool/sys_shd_common.py]: Done

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py (hunk 2 above)
Comment (RobertoAldea) [reply code/sys_shd/src/system_shared_tool/sys_shd_common.py]: Done

File: code/sys_shd/src/system_shared_tool/sys_shd_channels.py [hunk 6]
```diff
@@ -209,3 +209,16 @@ def is_empty(self) -> bool:
...
         return self.current_messages == 0
+
+    def terminate(self) -> None:
+        """Terminate the queue.
+        """
+        try:
+            self.close()
+            try:
+                self.unlink()
+            except ipc.ExistentialError as err:
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_channels.py]: merge into one try except case

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py [hunk 7]
```diff
@@ -19,6 +25,73 @@
...
+        """
+        super().__init__(group = None, target = node_params.target, name = node_params.name,
+                         args = node_params.args, kwargs = node_params.kwargs,
+                         daemon = node_params.daemon)
+        self.working_flag: Event = working_flag
+        self.cycle_period: float = cycle_period
+
+    def sync_shd_data(self) -> None:
+        """Function to be implemented in the inherited class
+        """
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_common.py:63]: Raise an exception `raise NotImplementedError()`

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py [hunk 8]
```diff
@@ -19,6 +25,73 @@
...
+
+    def process_iteration(self) -> None:
+        """Function to be implemented in the inherited class
+        """
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_common.py:68]: Raise exception

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py [hunk 9]
```diff
@@ -19,6 +25,73 @@
...
+
+    def stop(self) -> None:
+        """
+        Function to be implemented in the inherited class,
+        it will execute when finishing the thread.
+        """
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_common.py:75]: Raise exception

File: code/sys_shd/src/system_shared_tool/sys_shd_common.py [hunk 10]
```diff
@@ -19,6 +25,73 @@
...
+        Main method executed by the CAN thread. It receive data from EPCs and PLAKs
+        and store it on the corresponding chan.
+        '''
+        log.info("Start running process")
+        while self.working_flag.isSet():
+            try:
+                tic = time()
+                self.process_iteration()
+                end_tick = time()
+                waiting_ms = self.cycle_period-int(end_tick-tic)
```
Comment (mariuscrsn) [inline code/sys_shd/src/system_shared_tool/sys_shd_common.py]: Implement something like this: 
https://git.csic.es/EnergyLab/BFR/software/bms_bfr50kw/mid/MID_DABS/-/blob/develop/MID_DABS_Node.py?ref_type=heads#L390

Check sleeping time and if it is > 0, print an critical error

**Review by mariuscrsn — CHANGES_REQUESTED** [review]

File: code/sys_shd/src/system_shared_tool/sys_shd_channels.py (hunk 6 above)
Comment (RobertoAldea) [reply code/sys_shd/src/system_shared_tool/sys_shd_channels.py]: Done

**Review by mariuscrsn — APPROVED** [review]

**Review by mariuscrsn — APPROVED** [review]

**Pull request merged**